### PR TITLE
Spike: Add a zipper to mutate Ast (WIP)

### PR DESCRIPTION
**Pull request by andimarek**

### CODE REVIEW

File: src/main/java/graphql/language/ChildrenContainer.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+@PublicApi
+public class ChildrenContainer {
+
+    private final Map<String, List<Node>> children = new LinkedHashMap<>();
+
+    private ChildrenContainer(Map<String, List<Node>> children) {
+        this.children.putAll(Assert.assertNotNull(children));
+    }
+
+    public <T extends Node> List<T> getList(String key) {
```
Comment (bbakerman) [inline src/main/java/graphql/language/ChildrenContainer.java]: `getChildrenWithName` or just `getChildren` is a better namer I think.  You know its a List from the return type

File: src/main/java/graphql/language/AbstractNode.java [hunk 2]
```diff
@@ -47,4 +47,15 @@ public SourceLocation getSourceLocation() {
         return list.stream().map(Node::deepCopy).map(node -> (V) node).collect(Collectors.toList());
     }
 
+    //TODO: Implement really in each Node
+    @Override
+    public ChildrenContainer getNamedChildren() {
```
Comment (bbakerman) [inline src/main/java/graphql/language/AbstractNode.java]: make it abstract and they will have to!

File: src/main/java/graphql/language/AbstractNode.java [hunk 3]
```diff
@@ -47,4 +47,15 @@ public SourceLocation getSourceLocation() {
...
+        return null;
+    }
+
+    //TODO: Implement really in each Node
+    @Override
+    public T withNewChildren(ChildrenContainer newChildren) {
+        return null;
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/language/AbstractNode.java]: again make it abstract eventually to enforce this precondition

File: src/main/java/graphql/language/ChildrenContainer.java (hunk 1 above)
Comment (bbakerman) [reply src/main/java/graphql/language/ChildrenContainer.java]: still stand by this comment

File: src/main/java/graphql/language/ChildrenContainer.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return (List<T>) children.getOrDefault(key, new ArrayList<>());
+    }
+
+    public <T extends Node> T getSingleValueOrNull(String key) {
+        List<? extends Node> result = children.getOrDefault(key, new ArrayList<>());
+        if (result.size() > 1) {
+            throw new IllegalStateException("children " + key + " is not a single value");
+        }
+        return result.size() > 0 ? (T) result.get(0) : null;
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/language/ChildrenContainer.java]: getChildOrNull might ber a better name

File: src/main/java/graphql/language/ChildrenContainer.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/language/ChildrenContainer.java]: will change

File: src/main/java/graphql/language/AbstractNode.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/language/AbstractNode.java]: yes, it was just temp

File: src/main/java/graphql/language/AbstractNode.java (hunk 3 above)
Comment (andimarek) [reply src/main/java/graphql/language/AbstractNode.java]: yes, it was just temp

Comment (andimarek) [reply src/main/java/graphql/language/AbstractNode.java]: it is actually in Node

File: src/main/java/graphql/language/ChildrenContainer.java (hunk 4 above)
Comment (andimarek) [reply src/main/java/graphql/language/ChildrenContainer.java]: yes, will change

File: src/main/java/graphql/language/ObjectTypeDefinition.java [hunk 5]
```diff
@@ -72,10 +76,46 @@ public Description getDescription() {
...
+//            Node curChild = curChildren.get(ix);
+//            if(curChild instanceof Type ) {
+//               // this means it is a implementz child
+//               // but at which index in the implementz list?
+//            }else if( curChild instanceof Directive) {
+//                // this means it is a directives child
+//                // but at which index in the directives list
+//            }
+//        }
+//    }
```
[the remark this answers is not in the document]
Comment (andimarek) [reply src/main/java/graphql/language/ObjectTypeDefinition.java]: yes

**Comment by andimarek** [comment]
Work will continue here: https://github.com/graphql-java/graphql-java/pull/1335

**Pull request closed without merging**